app/scripts/directives/oscWebhookTriggers.js

```diff
@@ -0,0 +1,153 @@
+          $(deprecatedInputField).attr('type') === 'password' ? $(deprecatedInputField).attr('type', 'text') : $(deprecatedInputField).attr('type', 'password');
+        };
+
+        $scope.removeWebhookTrigger = function(index) {
+          if ($scope.webhookTriggers.length === 1) {
+            var lonelyTrigger = _.first($scope.webhookTriggers);
+            lonelyTrigger.lastTriggerType = "";
+            lonelyTrigger.data = {
+              type: ""
+            };
+          } else {
+            $scope.webhookTriggers.splice(index,1);
+          }
+          $scope.form.$setDirty();
+
+          // After removing trigger check check other triggers for duplicates.
+          _.each($scope.webhookTriggers, function(trigger) {
+            if (trigger.isDuplicit) {
+              duplicityCheck(trigger);
+            }
+          })
+        };
+
+        // Check if new or modified webhook trigger is a duplicate. If so, show a warning under the appropriate trigger
+        var duplicityCheck = function(trigger) {
+          var matchingTriggers = _.filter($scope.webhookTriggers, function(wehbookTrigger) {
+            return _.isEqual(wehbookTrigger.data, trigger.data);
+          });
+
+          if (matchingTriggers.length > 1) {
+            _.first(matchingTriggers).isDuplicit = false;
+            var matchingExceptFirst = _.takeRight(matchingTriggers, matchingTriggers.length - 1);
+            _.each(matchingExceptFirst, function(trigger) {
+              trigger.isDuplicit = true;
+            });
+            return
+          }
+          trigger.isDuplicit = false;
+        };
+
+        // When user changes the webhook trigger type move the secret data from the old one to the new one.
+        $scope.triggerTypeChange = function(trigger) {
+          var lastTriggerType = _.toLower(trigger.lastTriggerType);
+          var newTriggerType = _.toLower(trigger.data.type);
+
+          trigger.data = _.mapKeys(trigger.data, function(value, key) {
+            if (key === lastTriggerType) {
+              return newTriggerType;
+            }
+            return key;
+          });
+          trigger.lastTriggerType = trigger.data.type;
+          duplicityCheck(trigger);
+        };
+
+        $scope.triggerSecretChange = function(trigger) {
+          duplicityCheck(trigger);
+        };
+
+        var addWebhookTrigger = function() {
+          $scope.webhookTriggers.push({
+            lastTriggerType: "",
+            data: {
+              type: ""
+            }
+          });
+        };
+
+        // Check last trigger if it's type and secret are selected.
+        $scope.checkLastAndAddNew = function() {
+          var lastTrigger = _.last($scope.webhookTriggers);
+          var lastTriggerSecretData = $filter('getWebhookSecretData')(lastTrigger);
+          if (lastTrigger.data.type && (lastTriggerSecretData.secret || lastTriggerSecretData.secretReference.name)) {
```

Comment:
Isn't this a runtime error if secretReference isn't defined?

Reply:
The case shouldnt happen, since `lastTriggerSecretData.secret` or `lastTriggerSecretData.secretReference.name` should be set. But changing the check to `_.has()`